crates/chia-datalayer/src/merkle.rs

```diff
@@ -0,0 +1,1902 @@
+
+    // TODO: don't use String here
+    #[error("failed loading node: {0}")]
+    FailedLoadingNode(String),
+
+    #[error("blob length must be a multiple of block count, found extra bytes: {0}")]
+    InvalidBlobLength(usize),
+
+    #[error("key already present")]
+    KeyAlreadyPresent,
+
+    #[error("requested insertion at root but tree not empty")]
+    UnableToInsertAsRootOfNonEmptyTree,
+
+    #[error("unable to find a leaf")]
+    UnableToFindALeaf,
+
+    #[error("unknown key: {0:?}")]
+    UnknownKey(KvId),
+
+    #[error("key not in key to index cache: {0:?}")]
+    IntegrityKeyNotInCache(KvId),
+
+    #[error("zero-length seed bytes not allowed")]
+    ZeroLengthSeedNotAllowed,
+
+    #[error("block index out of range: {0:?}")]
+    BlockIndexOutOfRange(TreeIndex),
+
+    #[error("node not a leaf: {0:?}")]
+    NodeNotALeaf(InternalNode),
+
+    #[error("from streamable: {0:?}")]
+    Streaming(chia_traits::chia_error::Error),
+
+    #[error("index not a child: {0}")]
+    IndexIsNotAChild(TreeIndex),
+}
+
+// assumptions
+// - root is at index 0
+// - any case with no keys will have a zero length blob
+
+// define the serialized block format
+const METADATA_RANGE: Range<usize> = 0..METADATA_SIZE;
+const METADATA_SIZE: usize = 2;
+// TODO: figure out the real max better than trial and error?
+const DATA_SIZE: usize = 53;
+const BLOCK_SIZE: usize = METADATA_SIZE + DATA_SIZE;
+type BlockBytes = [u8; BLOCK_SIZE];
+type MetadataBytes = [u8; METADATA_SIZE];
+type DataBytes = [u8; DATA_SIZE];
+const DATA_RANGE: Range<usize> = METADATA_SIZE..METADATA_SIZE + DATA_SIZE;
+
+fn streamable_from_bytes_ignore_extra_bytes<T>(bytes: &[u8]) -> Result<T, Error>
+where
+    T: Streamable,
+{
+    let mut cursor = std::io::Cursor::new(bytes);
+    // TODO: consider trusted mode?
+    T::parse::<false>(&mut cursor).map_err(Error::Streaming)
+}
+
+#[repr(u8)]
+#[derive(Clone, Copy, Debug, Hash, Eq, PartialEq, Streamable)]
+pub enum NodeType {
+    Internal = 0,
+    Leaf = 1,
+}
+
+impl NodeType {
+    pub fn from_u8(value: u8) -> Result<Self, Error> {
+        streamable_from_bytes_ignore_extra_bytes(&[value])
+    }
+
+    #[allow(clippy::wrong_self_convention, clippy::trivially_copy_pass_by_ref)]
+    pub fn to_u8(&self) -> u8 {
+        Streamable::to_bytes(self).unwrap()[0]
+    }
+}
+
+#[allow(clippy::needless_pass_by_value)]
+fn sha256_num<T: ToBytes>(input: T) -> Hash {
+    let mut hasher = Sha256::new();
+    hasher.update(input.to_be_bytes());
+
+    Bytes32::new(hasher.finalize())
+}
+
+fn sha256_bytes(input: &[u8]) -> Hash {
+    let mut hasher = Sha256::new();
+    hasher.update(input);
+
+    Bytes32::new(hasher.finalize())
+}
+
+fn internal_hash(left_hash: &Hash, right_hash: &Hash) -> Hash {
+    let mut hasher = Sha256::new();
+    hasher.update(b"\x02");
+    hasher.update(left_hash);
+    hasher.update(right_hash);
+
+    Bytes32::new(hasher.finalize())
+}
+
+#[cfg_attr(feature = "py-bindings", pyclass(name = "Side", eq, eq_int))]
```

Comment:
You shouldn't need to set the name here since it's going to default to the struct name.